internal/infrastructure/price-feeder/service.go

```diff
@@ -0,0 +1,265 @@
+package pricefeederinfra
+
+import (
+	"context"
+	"sync"
+
+	pricefeeder "github.com/tdex-network/tdex-daemon/pkg/price-feeder"
+
+	"github.com/tdex-network/tdex-daemon/internal/core/ports"
+)
+
+var (
+	sources = map[string]struct{}{
+		"kraken": {},
+	}
+)
+
+type priceFeederService struct {
+	feederSvc           pricefeeder.PriceFeeder
+	priceFeedRepository PriceFeedRepository
+	feedChan            chan ports.PriceFeedChan
+
+	marketsMtx sync.RWMutex
+	markets    []pricefeeder.Market
+}
+
+func NewService(
+	feederSvc pricefeeder.PriceFeeder,
+	priceFeedRepository PriceFeedRepository,
+) ports.PriceFeeder {
+	return &priceFeederService{
+		feederSvc:           feederSvc,
+		priceFeedRepository: priceFeedRepository,
+		feedChan:            make(chan ports.PriceFeedChan),
+		marketsMtx:          sync.RWMutex{},
+		markets:             make([]pricefeeder.Market, 0),
+	}
+}
+
+func (p *priceFeederService) Start(
+	ctx context.Context,
+	markets []ports.Market,
+) (chan ports.PriceFeedChan, error) {
+	mkts := make([]pricefeeder.Market, len(markets))
+	for _, v := range markets {
+		priceFeed, err := p.priceFeedRepository.GetPriceFeedsByMarket(ctx, Market{
+			BaseAsset:  v.GetBaseAsset(),
+			QuoteAsset: v.GetQuoteAsset(),
+		})
+		if err != nil {
+			return nil, err
+		}
+
+		mkts = append(mkts, pricefeeder.Market{
+			BaseAsset:  v.GetBaseAsset(),
+			QuoteAsset: v.GetQuoteAsset(),
+			Ticker:     priceFeed.Market.Ticker,
+		})
+	}
+
+	if err := p.watchMarkets(mkts); err != nil {
+		return nil, err
+	}
+
+	return p.feedChan, nil
+}
+
+func (p *priceFeederService) watchMarkets(
+	markets []pricefeeder.Market,
+) error {
+	if err := p.feederSvc.SubscribeMarkets(markets); err != nil {
+		return err
+	}
+
+	if err := p.feederSvc.Start(); err != nil {
+		return err
+	}
+
+	go func() {
+		for priceFeed := range p.feederSvc.FeedChan() {
+			p.feedChan <- Feed{
+				Market: Market{
+					BaseAsset:  priceFeed.Market.BaseAsset,
+					QuoteAsset: priceFeed.Market.QuoteAsset,
+				},
+				Price: Price{
+					BasePrice:  priceFeed.Price.BasePrice,
+					QuotePrice: priceFeed.Price.QuotePrice,
+				},
+			}
+		}
+	}()
+
+	return nil
+}
+
+func (p *priceFeederService) Stop(ctx context.Context) {
+	p.feederSvc.Stop()
+	close(p.feedChan)
+}
+
+func (p *priceFeederService) StartFeed(ctx context.Context, feedID string) error {
+	p.feederSvc.Stop()
+
+	priceFeed, err := p.priceFeedRepository.GetPriceFeed(ctx, feedID)
+	if err != nil {
+		return err
+	}
+
+	p.addMarkets([]pricefeeder.Market{
+		{
+			BaseAsset:  priceFeed.Market.BaseAsset,
+			QuoteAsset: priceFeed.Market.QuoteAsset,
+			Ticker:     priceFeed.Market.Ticker,
+		},
+	})
+
+	return p.watchMarkets(p.getMarkets())
```

Comment:
I guess this comes from the old implementation where we had a different approach and start all price feeds for all markets defined in a source config file.
Now we are shifting to a totally different one, so here theoretically we should interact wit the price source and start receiving feeds from it that will be forwarded to a channel.. on the other side of this channel there will be the app service that will update markets prices.

Reply:
Done